### ISSUE TITLE
Customizable list of file extension to check

### ISSUE THREAD

**Issue by svogt** [versioneye/veye-checker#18]
Would be nice if the list of extensions to send to Versioneye could be configurable (to also include some more exotic extensions like .war, .ear, etc.)

**Comment by timgluz** [comment]
Hi,
tl;dr- great feature, must think and discuss a bit as may require some changes in the overall logic;

Current solution makes naive expectation that all the binaries are released with s.o "official" file-extension, so we would know which digest algorithm to use and  we could match the result with an already existing data.

I have a solution for this issue -  using an unified digest algorithm (e.g [Blake2](https://en.wikipedia.org/wiki/BLAKE_%28hash_function%29) )  for all the binaries and then we could use white&blacklist for file-extensions. I really like this idea, but it requires extra service which download binaries and calculates digest for each of them. Technically it's simple, but it will consume extra computing power and financial resources.

But it could be also done by adding a `register` command that calculates a sha-value and asks  user some product details and then saves the data into the database - aka crowd-sourcing this big tasks to users.

I'll think a bit more about this.

**Comment by reiz** [comment]
@timgluz I think what @svogt means is that the file extensions for which we generate SHAs locally are configurable. Currently we have SHAs for files from Maven Central, Nuget, NPMRegistry and PyPi. But Maven Central is not just hosting JAR files. We have for example 17631 WAR files and 311 EAR files from Maven Central. 

Another use could be that you have a directory with JARs, WARs and EARs but you want to scan only the WARs for example. 

We could have a file extension filter in the `veye_checker.toml` to control which file types are processed locally.

**Comment by timgluz** [comment]
Hi,

i thought about that issue - i will flip the current implementation around, so that the tool will not use anymore hardcoded s.o official file-extension to dispatch digest algorithms, but will use a table with structure `{"sha-algo" => [list of file extensions]}` that way an user can control which digest algorithms to use and which file-extensions belong to what digest algorithm. 

By default this table will be populated with official file-extensions to keep app behave like it is now.
Here're a few usage examples:

1. This example will scan only `ear,war and jar` files and will use only `md5, sha1` algos
```
veye-checker shas ../bins --use-digests="md5,sha1" --ext-md5="ear,war" --ext-sha1="jar"
```

2. This example will use default digests algos, but will overwrite file extensions for sha512:

```
veye-checker shas ../nuget --ext-sha512="nupkg,nuzip"
```

3. This example will not scan any `md5 or sha1` packages:

```
veye-checker shas ../bins --no-digests="md5,sha1"
```

4. Empty extension will equal to "--no-digests"

```
veye-checker shas ../bins --ext-md5=""
```

The same file-extension may belong to many digest-group, so user can check same file with multiple digest algorithm.

As it's complicated configuration, i'll skip environment variables for that and only use configs from command line and from `veye_checker.toml`

it's a big refactoring and will take a couple of days.

I think this task needs also a extra command `show-exts`, which could print out the configuration of digest algorithms and file-extensions.

**Comment by reiz** [comment]
@timgluz keep it simple. For right now just add something linke that to the `veye_checker.toml`: 

```
[config]
file_extension_filter = "war,ear,aar"
```
With the above config the tool should only handle files which end on `war` or `ear` or `aar` and all other file types are skipped with that config. @svogt What do you think? Good enough for now?

**Comment by timgluz** [comment]
I've done with re-writing the core logic of dispatching digest algos - it's now dynamic and allows to specify which file digest algorithm to use for each file-extension and ables to ignore some algorithms;

There's still a few things to do before it'll be merged into `master` branch, specifically:

* parsing command-line arguments to get values for the `DigestExtTable`
* filling a table of DigestExtensions from configs
* update tasks to handle multiple shas from digest process and api;

But i'll come back to this issue when other more urgent tasks are done;

**Comment by svogt** [comment]
Sorry for the late response. For me just the inclusion filter is sufficient, but i think I see @timgluz point as you need to figure out which digest to use. Only from the binary (extension) you do not know to which package manager it belongs. Of course for WAR and EAR it's pretty easy, those are only used in the Java ecosystem, so the Maven digests are the only ones that make sense. For others it might not be that easy.

Having a key=value config or a per digest config (--md5=zip,tar.gz,tar.bz2 --sha1=war,jar,ear) is also fine. That per digest thing should also be pretty easy to get via env variables as you can have a dedicated variable per digest and use a comma separated list of extensions to scan for.

**Comment by timgluz** [comment]
Done, 

user can now block use of digest algorithm and overwrite file-extensions for each of digest algorithms.

here's example:

```
veye_checker shas tests/fixtures/files/ --no-md5 --ext-sha1="whl,jar"
```
